library/Odesk/Phystrix/AbstractCommand.php

```diff
@@ -219,33 +219,32 @@ public function execute()
         if ($this->isRequestCacheEnabled()) {
             $fromCache = $this->requestCache->get($this->getCommandKey(), $this->getCacheKey());
             if ($fromCache !== null) {
-                $this->executionEvents[] = self::EVENT_RESPONSE_FROM_CACHE;
                 $metrics->markResponseFromCache();
+                $this->handleEvent(self::EVENT_RESPONSE_FROM_CACHE);
                 return $fromCache;
             }
         }
         $circuitBreaker = $this->getCircuitBreaker();
         if (!$circuitBreaker->allowRequest()) {
-            $this->executionEvents[] = self::EVENT_SHORT_CIRCUITED;
             $metrics->markShortCircuited();
+            $this->handleEvent(self::EVENT_SHORT_CIRCUITED);
             return $this->getFallbackOrThrowException();
         }
         $this->invocationStartTime = $this->getTimeInMilliseconds();
         try {
             $result = $this->run();
             $this->recordExecutionTime();
-            $this->executionEvents[] = self::EVENT_SUCCESS;
             $metrics->markSuccess();
             $circuitBreaker->markSuccess();
+            $this->handleEvent(self::EVENT_SUCCESS);
         } catch (BadRequestException $exception) {
             // Treated differently and allowed to propagate without any stats tracking or fallback logic
             $this->recordExecutionTime();
             throw $exception;
         } catch (Exception $exception) {
             $this->recordExecutionTime();
-            $this->executionException = $exception;
-            $this->executionEvents[] = self::EVENT_FAILURE;
             $metrics->markFailure();
+            $this->handleEvent(self::EVENT_FAILURE, $exception);
```

Comment:
@lcf I agree:
1. let's make handleEvent private, will serve as centralized function for performing the list addition, exception set (as per comment above), 
2. after, handleEvent can call either a protected function that can be overridden, or a registration-based closure-type assembly for downstream customization. 

#2 is super important for me, as phystrix does not have time-limiting or segfault-bypassing powers, so if the request times out, so does logging that it ever happened. We've deployed our fork (including all my PRs) to prod, and attached our symfony event system to use our logging system to record events as they happen. When requests time out completely, we can pinpoint where and how.